src/vsr/replica.zig

```diff
@@ -7659,6 +7610,92 @@ pub fn ReplicaType(
+                (self.status == .recovering and self.log_view == self.view) or
+                (self.status == .view_change and self.log_view == self.view));
+
+            const primary_repairing =
+                self.status == .view_change and self.log_view == self.view;
+            if (primary_repairing) {
+                assert(self.primary_index(self.view) == self.replica);
+                assert(self.do_view_change_quorum);
+            }
+
+            assert(self.view == self.log_view);
+
+            // The DVC headers include:
+            // - all available cluster-uncommitted ops, and
+            // - the highest cluster-committed op (if available).
+            // We cannot safely go beyond that in all cases for fear of concealing a break:
+            // - During a prior view-change we might have only accepted a single header from the
+            //   DVC: "header.op = op_prepare_max", and then not completed any
+            //   repair.
+            // - Similarly, we might have receive a catch-up SV message and only installed a
+            //   single (checkpoint trigger) hook header.
+            //
+            // DVC headers are stitched together from the journal and existing view headers (they
+            // might belong to the next log wrap), to guarantee that a DVC with log_view=v includes
+            // all uncommitted ops with views <v. Special case: a primary that has collected a DVC
+            // quorum and installed surviving headers into the journal ignores view headers (they
+            // might have some truncated ops).
+            var view_headers_updated = vsr.Headers.ViewChangeArray{
+                .command = .do_view_change,
+                .array = .{},
+            };
+
+            const view_headers_op_max = self.view_headers.array.get(0).op;
+            var op = if (primary_repairing) self.op else @max(self.op, view_headers_op_max);
+            for (0..constants.pipeline_prepare_queue_max + 1) |_| {
+                const header_journal: ?*const vsr.Header.Prepare =
+                    self.journal.header_with_op(op);
+
+                const header_view: ?*const vsr.Header.Prepare = header: {
+                    if (primary_repairing or op > view_headers_op_max) break :header null;
+
+                    const header = &self.view_headers.array.const_slice()[view_headers_op_max - op];
+                    break :header switch (vsr.Headers.dvc_header_type(header)) {
+                        .valid => header,
+                        .blank => null,
+                    };
+                };
+
+                if (header_journal != null and header_view != null) {
+                    assert(header_journal.?.op == header_view.?.op);
+                    assert(header_journal.?.view == header_view.?.view);
+                    assert(header_journal.?.checksum == header_view.?.checksum);
+                }
+
+                if (header_journal == null and header_view == null) {
```

Comment:
- `assert(view_headers_updated.array.len > 0);`: The new "op head" is never a blank.
- `assert(op != self.op);`: `self.op`'s header always exists in journal.